util/ipgen/renderer.py

```diff
@@ -0,0 +1,263 @@
+                "Output staging directory '{}' already exists. Remove it and "
+                "try again.".format(output_dir_staging))
+
+        template_path = self.ip_template.template_path
+
+        try:
+            # Copy everything but the templates.
+            shutil.copytree(template_path,
+                            output_dir_staging,
+                            ignore=shutil.ignore_patterns('*.tpl'))
+
+            # Render templates.
+            for template_filepath in template_path.glob('**/*.tpl'):
+                template_filepath_rel = template_filepath.relative_to(
+                    template_path)
+
+                # Put the output file into the same relative directory as the
+                # template. The output file will also have the same name as the
+                # template, just without the '.tpl' suffix.
+                outdir_path = output_dir_staging / template_filepath_rel.parent
+
+                self._render_mako_template_to_file(template_filepath,
+                                                   outdir_path)
+
+            # Generate register interface through reggen.
+            hjson_path = (output_dir_staging / 'data' /
+                          (self.ip_template.name + '.hjson'))
+            rtl_path = output_dir_staging / 'rtl'
+            # TODO: Pass on template parameters to reggen? Or enable the user
+            # to set a different set of parameters in the renderer?
+            reggen.gen_rtl.gen_rtl(IpBlock.from_path(str(hjson_path), []),
+                                   str(rtl_path))
+
+            # Write IP configuration (to reproduce the generation process).
+            # TODO: Should the ipconfig file be written to the instance name,
+            # or the template name?
+            self.ip_config.to_file(
+                output_dir_staging /
+                'data/{}.ipconfig.hjson'.format(self.ip_config.instance_name))
+
+            # Safely overwrite the existing directory if necessary:
+            #
+            # - First move the existing directory out of the way.
+            # - Then move the staging directory with the new content in place.
+            # - Finally remove the old directory.
+            #
+            # If anything goes wrong in the meantime we are left with either
+            # the old or the new directory, and potentially some backups of
+            # outdated files.
+            do_overwrite = overwrite_output_dir and output_dir.exists()
+            output_dir_existing_bak = output_dir.with_suffix(
+                '.bak~' + str(int(time.time())))
+            if do_overwrite:
+                os.rename(output_dir, output_dir_existing_bak)
+
+            # Move the staging directory to the final destination.
+            os.rename(output_dir_staging, output_dir)
+
+            # Remove the old/"overwritten" data.
+            if do_overwrite:
+                shutil.rmtree(output_dir_existing_bak, ignore_errors=True)
```

Comment:
I know it shouldn't happen, but would it make sense to check whether the backup directory was removed and, if not, write something to stderr? It might be helpful for a user.

Reply:
Good point, I removed the error silencing and am now re-throwing the exception with a bit more info (which will lead to a log print ultimately). Unfortunately (as so often in Python), shutil.rmtree() isn't really specific about what exceptions one can expect, so it's a bit of a catch-all now.